presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -682,6 +686,187 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
+            return encodeUnscaledValue(0);
+        }
+        if (inputScale == 0) {
+            return encodeUnscaledValue(num);
+        }
+        if (num.signum() < 0) {
+            return encodeUnscaledValue(roundLongLongDecimal(num.negate(), rescaleFactor).negate());
+        }
+        else {
+            return encodeUnscaledValue(roundLongLongDecimal(num, rescaleFactor));
+        }
+    }
+
+    @UsedByGeneratedCode
+    private static BigInteger roundLongLongDecimal(BigInteger num, BigInteger rescaleFactor)
+    {
+        BigInteger[] divideAndRemainder = num.divideAndRemainder(rescaleFactor);
+        BigInteger roundUp = divideAndRemainder[1].compareTo(rescaleFactor.shiftRight(1)) >= 0 ? ONE : ZERO;
+        return divideAndRemainder[0].add(roundUp);
+    }
+
+    @UsedByGeneratedCode
+    public static long roundLongShortDecimal(Slice numSlice, BigInteger rescaleFactor, long inputScale)
+    {
+        BigInteger num = decodeUnscaledValue(numSlice);
+        if (num.signum() == 0) {
+            return 0;
+        }
+        if (num.signum() < 0) {
+            return roundLongLongDecimal(num.negate(), rescaleFactor).negate().longValue();
+        }
+        else {
+            return roundLongLongDecimal(num, rescaleFactor).longValue();
+        }
+    }
+
+    private static SqlScalarFunction decimalRoundNFunction()
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .name("round")
+                .longVariableConstraints(
+                        //result precision = increment the input precision only if the input number has a decimal point (scale > 0)
+                        longVariableExpression("return_precision", "min(38, num_precision + 1)"))
+                .argumentTypes(parseTypeSignature("decimal(num_precision,num_scale)", ImmutableSet.of("num_precision", "num_scale")), parseTypeSignature(StandardTypes.BIGINT))
+                .returnType(parseTypeSignature("decimal(return_precision,num_scale)", ImmutableSet.of("return_precision", "num_scale")))
+                .build();
+        return SqlScalarFunction.builder(MathFunctions.class)
+                .signature(signature)
+                .description("round to given number of decimal places")
+                .implementation(b -> b
+                        .methods("roundNShortShortDecimal", "roundNLongLongDecimal", "roundNShortLongDecimal")
+                        .withExtraParameters(MathFunctions::decimalRoundNExtraParameters))
+                .build();
+    }
+
+    private static List<Object> decimalRoundNExtraParameters(SpecializeContext context)
+    {
+        return ImmutableList.of(context.getLiteral("num_precision"), context.getLiteral("num_scale"));
+    }
+
+    @UsedByGeneratedCode
+    public static long roundNShortShortDecimal(long num, long roundScale, long inputPrecision, long inputScale)
+    {
+        if (num == 0 || inputPrecision - inputScale + roundScale <= 0) {
```

Comment:
It is actually checked in `private static Slice roundNLongDecimal` called by both.